test/unit/master/test_build.py

```diff
@@ -375,6 +375,32 @@ def test_canceling_build_sets_canceled_timestamp(self):
         self.assertIsNotNone(build.get_state_timestamp(BuildStatus.CANCELED),
                              '"canceled" timestamp should be set when build is canceled.')
 
+    def test_get_failed_atoms_returns_none_if_not_finished(self):
+        build = self._create_test_build(BuildStatus.BUILDING)
+        self.assertIsNone(build._get_failed_atoms())
+
+    def test_get_failed_atoms_returns_empty_list_if_finished_and_all_passed(self):
+        build = self._create_test_build(BuildStatus.FINISHED)
+        build._build_artifact = MagicMock(spec_set=BuildArtifact)
+        build._build_artifact.get_failed_subjob_and_atom_ids.return_value = []
+
+        self.assertEquals([], build._get_failed_atoms())
+
+    def test_get_failed_atoms_returns_fails_atoms_only(self):
+        build = self._create_test_build(BuildStatus.FINISHED)
+        build._build_artifact = MagicMock(spec_set=BuildArtifact)
+        # Failed items: (SubjobId: 1, AtomId: 1) and (SubjobId: 3, AtomId: 3)
+        build._build_artifact.get_failed_subjob_and_atom_ids.return_value = [(1, 1), (3, 3)]
+        all_subjobs = self._create_subjobs(count=5, num_atoms_each=10)
```

Comment:
I think you might be able to use the `num_subjobs` and `num_atoms_per_subjob` args to `_create_test_build()` and it will set up the build with those subjobs for you.

Man I forgot how insane the tests for this class are.